### PR TITLE
DOC: getting started with jax for AI

### PR DESCRIPTION
**Pull request by jakevdp**
Initial draft of the first section of a JAX intro for ML/AI-focused users.

### CODE REVIEW

**Comment by pavithraes** [comment]
This tutorial has a very good pedagogical flow, the concepts were clear and I was able to follow along well. :)

Since this project used `myst-nb`, and inline comments are not suported in the rich-diff mode for Jupyter Notebooks, I've incorporated some suggestions in my fork, [link to PR preview for the diff](https://github.com/jax-ml/jax-ai-stack/compare/main...pavithraes:review-get-started-jax?expand=1#diff-bfadadb94f7928eb67c9aec296095df29fe6aa3d6f640af15b4ee5c56659d25e).

Overview of the suggestions:
- I understand, this is an introduction to the pillars of the jax-ai-stack with an emphasis on JAX because it's at the heart of the stack. As someone not familiar with the ecosystem, I was initially a bit uncertain about the learning outcomes, so I've suggested some updates to the introductory sections (and a couple of headings) to add some more clarity to this purpose.
- Almost all other updates are minor: style-related (mainly following project-specific capitalization standards), wording and links.

---

This is not in my branch, but I noticed the mention of `nnx.grad` in the narrative, but the example itself uses `nnx.value_and_grad`. Perhaps we can update the example or clarify in the narration?

Mentions:

> Training the model section:
> Notice here the use of `nnx.jit` and `nnx.grad`

> Automatic differentiation: 
> In the example above, we used the similar `nnx.grad` function, which has special handling for `flax.nnx` objects.

---

Aside, would it be feasible to configure [ReviewNB](https://www.reviewnb.com/) for this repository? I think it can be helpful and more convienent for inline-reviews in notebooks. If we do, I'll be happy to submit my thoughts here as inline-review-comments.

File: docs/getting_started_with_jax_for_AI.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    data: jax.Array,
+    labels: jax.Array):
+  loss_gradient = nnx.value_and_grad(loss_fun, has_aux=True)  # gradient transform!
+  (loss, logits), grads = loss_gradient(model, data, labels)
+  optimizer.update(grads)  # inplace update
+```
+
++++ {"id": "K2Tp-ym6sXEl"}
+
+Notice here the use of `nnx.jit` and `nnx.grad`, which are transformations built on `jax.jit` and `jax.grad`: `jit` is a Just In Time compilation transformation, and will cause the function to be passed to the XLA compiler for fast repeated execution. `grad` is a gradient transformation, and uses JAX's automatic differentiation for fast optimization of large networks.
```
Comment (caraya10) [inline docs/getting_started_with_jax_for_AI.md:172]: One thing we discussed is the right place to talk about debugging and best practices for how to do debugging of jitted functions.  Since we just started using JIT here we should discuss some simple options briefly and then link to where it is discussed in more depth.

File: docs/getting_started_with_jax_for_AI.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+                         gridspec_kw=dict(hspace=0.1, wspace=0.1))
+
+for i, ax in enumerate(axes.flat):
+    ax.imshow(digits.images[i], cmap='binary', interpolation='gaussian')
+    ax.text(0.05, 0.05, str(digits.target[i]), transform=ax.transAxes, color='green')
+```
+
++++ {"id": "Z3l45KgtfUUo"}
+
+Let's split these into a training and testing set, and convert these splits into JAX arrays which will be ready to feed into our model:
```
[the remark this answers is not in the document]
Comment (caraya10) [reply docs/getting_started_with_jax_for_AI.md:83]: Also, I see that below we talk about the speed up that JIT gives you along with timing below.  Should we consider doing that more in context here?

File: docs/getting_started_with_jax_for_AI.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+(selu(x + eps) - selu(x)) / eps
+```
+
++++ {"id": "pkQW2Hd_bPSd"}
+
+Importantly, the automatic differentiation approach is both more accurate and more efficient than computing numerical gradients. JAX's documentation has more discussion of autodiff at [Automatic differentiation](https://jax.readthedocs.io/en/latest/automatic-differentiation.html).
+
++++ {"id": "xsKyfRDNbj2y"}
+
+### Automatic vectorization
```
Comment (caraya10) [inline docs/getting_started_with_jax_for_AI.md:358]: I really like how the explanation is done here!

File: docs/getting_started_with_jax_for_AI.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+- [optax](http://optax.readthedocs.io): a tool designed for high-performance
+  optimization of functions in JAX, including the loss functions used in neural network training.
+
+Once you've worked through this content, you may wish to visit http://jax.readthedocs.io/ for a deeper dive into the JAX library itself.
+
++++ {"id": "z7sAr0sderhh"}
+
+## Example: a simple neural network with flax
+
+We'll start with a very quick example of what it looks like to use JAX with the [flax](https://flax.readthedocs.io) framework to define and train a very simple neural network on a hand-written digits dataset.
```
Comment (caraya10) [inline docs/getting_started_with_jax_for_AI.md:43]: Nit:  define and train a very simple neural network to recognize hand-written digits.

Essentially, what is the pragmatic goal of the mode.

File: docs/getting_started_with_jax_for_AI.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+import jax.numpy as jnp
+images_train, images_test, label_train, label_test = map(jnp.asarray, splits)
+print(f"{images_train.shape=} {label_train.shape=}")
+print(f"{images_test.shape=}  {label_test.shape=}")
+```
+
++++ {"id": "JzrixENjifiq"}
+
+### Defining the flax model
```
Comment (caraya10) [inline docs/getting_started_with_jax_for_AI.md:104]: So this is the first time the user is seeing rngs.  Should we add a sentence about them why they are important?

File: docs/getting_started_with_jax_for_AI.md (hunk 4 above)
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:43]: Also:

```
Let's start with a very simple example of using JAX with [Flax](https://flax.readthedocs.io)
to define a model and train it on the hand-written digits dataset with the help of
[Optax](https://optax.readthedocs.io) for optimization during training.
```

File: docs/getting_started_with_jax_for_AI.md (hunk 5 above)
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:104]: Good idea. Check this out:

```
**Note:** The [`flax.nnx.Module`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/module.html#flax.nnx.Module)
system is covered in more detail in the
[Flax basics](https://flax.readthedocs.io/en/latest/nnx_basics.html#the-flax-nnx-module-system) tutorial. Notice the use of the
[`flax.nnx.Rngs`](https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/rnglib.html#flax.nnx.Rngs) 
type - this is the main convenience API in Flax for managing random state.
The `Rngs` object can be can be used to get new unique keys based on a root JAX 
[pseudorandom number generator (PRNG)](https://jax.readthedocs.io/en/latest/random-numbers.html)
key passed to the constructor. Learn more about PRNGs in the Flax
[Randomness](https://flax.readthedocs.io/en/latest/guides/randomness.html)
and the JAX [PRNG](https://jax.readthedocs.io/en/latest/random-numbers.html)
tutorials.
```

cc @cgarciae

Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:104]: https://github.com/jax-ml/jax-ai-stack/pull/54

File: docs/getting_started_with_jax_for_AI.md (hunk 4 above)
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:43]: https://github.com/jax-ml/jax-ai-stack/pull/54

File: docs/getting_started_with_jax_for_AI.md [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+# Getting started with JAX
+
+[JAX](http://jax.readthedocs.io) is a Python package for accelerator-oriented array computation and program transformation, and is the engine behind cutting-edge AI research and production models at Google and beyond.
+
++++ {"id": "lN1DEDeMel9r"}
+
+## Who is this tutorial for?
+
+This tutorial is for those who want to get started using JAX to build and train neural network models. It assumes some familiarity with numerical computing in Python with [NumPy](http://numpy.org), and assumes some conceptual familiarity with defining, training, and evaluating machine learning models.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:24]: @caraya10 @jakevdp 

WDYT about 

```
# Getting started with JAX for AI

In this tutorial, you will learn how to get started using [JAX](https://jax.readthedocs.io) and JAX-based libraries to build and train a simple neural network model. JAX is a Python package for hardware accelerator-oriented array computation and program transformation, and is the engine behind cutting-edge AI research and production models at Google and beyond.

The JAX API focuses on [array-based](https://jax.readthedocs.io/en/latest/key-concepts.html#jax-arrays-jax-array) computation, and is at the core of a growing [ecosystem](https://jax.readthedocs.io/en/latest/index.html#ecosystem) of various domain-specific tools. This tutorial introduces a part of that ecosystem with a focus on AI, namely:

- [Flax](https://flax.readthedocs.io): A machine learning library designed for building and training scalable neural networks in JAX.
- [Optax](https://optax.readthedocs.io): A high-performance function optimization library that comes with built-in optimizers and loss functions. It also allows you to create your own such functions.
```

https://github.com/jax-ml/jax-ai-stack/pull/54 (feel free to comment, propose edits). Sorry for duplication

File: docs/getting_started_with_jax_for_AI.md [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+
+## Example: a simple neural network with flax
+
+We'll start with a very quick example of what it looks like to use JAX with the [flax](https://flax.readthedocs.io) framework to define and train a very simple neural network on a hand-written digits dataset.
+
++++ {"id": "pOlnhK-EioSk"}
+
+### Loading the data
+
+JAX can work with a variety of data loaders, but for simplicity here we can use the well-known [scikit-learn `digits`](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_digits.html) dataset.
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:49]: @caraya10 @jakevdp 

WDYT about 

```
JAX can work with a variety of data loaders, including [Grain](https://github.com/google/grain), [TensorFlow Datasets](https://github.com/tensorflow/datasets) and [TorchData](https://github.com/pytorch/data), but for simplicity
```

until the Dataloaders page is ready. Flax Linen worked with TFDS and PyTorch (noticed it's called TochData now).

Again, sorry for duplication https://github.com/jax-ml/jax-ai-stack/pull/54/

File: docs/getting_started_with_jax_for_AI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/getting_started_with_jax_for_AI.md:83]: Added `jax.Array` docs https://github.com/jax-ml/jax-ai-stack/pull/54/

Also added links to both JAX "Key concepts" (covering NumPy) and "Quickstart" docs throughout the Getting Started.

LMKWYT @caraya10 @jakevdp 

```
Next, split the dataset into a training and testing set, and convert these splits into [`jax.Array`s]https://jax.readthedocs.io/en/latest/key-concepts.html#jax-arrays-jax-array) before you feed them into the model.

> **Note:** A [`jax.Array`](https://jax.readthedocs.io/en/latest/_autosummary/jax.Array.html#jax.Array) is similar to [`numpy.ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray). You can learn more about `jax.Array`s in the [Key concepts](https://jax.readthedocs.io/en/latest/key-concepts.html#jax-arrays-jax-array) doc on the [JAX documentation site](https://jax.readthedocs.io/).
```

File: docs/getting_started_with_jax_for_AI.md (hunk 5 above)
Comment (jakevdp) [reply docs/getting_started_with_jax_for_AI.md:104]: This is also the first time seeing several other things as well (e.g. `nnx.Module`, `nnx.Linear`, etc.) I'm not sure how to explain all of those things without distracting from the main flow. I think our users are smart: if they don't understand something here, they can google it.

File: docs/getting_started_with_jax_for_AI.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/getting_started_with_jax_for_AI.md:49]: Once the dataloaders page is ready, we can reference it here. Until then, I don't think it's useful to have filler text or to link to specific projects.

File: docs/getting_started_with_jax_for_AI.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakevdp) [reply docs/getting_started_with_jax_for_AI.md:83]: Done in #64

File: docs/getting_started_with_jax_for_AI.md (hunk 5 above)
Comment (jakevdp) [reply docs/getting_started_with_jax_for_AI.md:104]: The context here: our initial attempt at JAX documentation tried to explain everything from the ground up, so that each time a user sees a concept it's fully explained: https://jax.readthedocs.io/en/latest/tutorials.html

The main feedback from that is that it's too boring and expository. So we're creating this tutorial flow as a response. For that reason, we're *explicitly* choosing not to explain every concept as it comes up, because we already have a doc flow that does that, and people don't like it.
